fleetspeak/src/server/internal/cache/cache.go

```diff
@@ -0,0 +1,114 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/google/fleetspeak/fleetspeak/src/common"
+	"github.com/google/fleetspeak/fleetspeak/src/server/db"
+)
+
+// Controls cache timing behavior, variable only to support unit tests.
+var (
+	maxAge         = 30 * time.Second
+	expireInterval = 5 * time.Minute
+)
+
+// Client is a cache of recently connected clients.
+type Client struct {
+	m    map[common.ClientID]*clientEntry
+	l    sync.RWMutex
+	stop chan struct{}
+}
+
+type clientEntry struct {
+	u time.Time
+	d *db.ClientData
+}
+
+// NewClient returns a new cache of client data.
+func NewClient() *Client {
+	ret := &Client{
+		m:    make(map[common.ClientID]*clientEntry),
+		stop: make(chan struct{}),
+	}
+	go ret.expireLoop()
+	return ret
+}
+
+// Get returns the cached client data, if there is sufficiently fresh data in
+// the cache, otherwise nil.
+func (c *Client) Get(id common.ClientID) *db.ClientData {
+	c.l.RLock()
+	defer c.l.RUnlock()
+
+	e := c.m[id]
+	if e == nil || db.Now().Sub(e.u) > maxAge {
+		return nil
```

Comment:
```
if e == nil {
  return nil
}
if db.Now().Sub(e.u) > maxAge {
  delete(c.m, id)
  return nil
}
```

Reply:
We only have a read lock at this point, so we cannot change the map.